python/lsst/ts/standardscripts/scheduler/base_set_desired_state.py

```diff
@@ -0,0 +1,214 @@
+            else salobj.State(current_summary_state.summaryState)
+        )
+
+    async def handle_no_summary_state_data(self) -> None:
+        """Handle condition where no information about Scheduler summary state
+        is available.
+
+        Start by assuming Scheduler is in STANDBY, if this fails, assume it is
+        in DISABLED and finally in ENABLED.
+        """
+
+        for coro in self._tasks_to_handle_summary_state:
+            if await coro():
+                return
+
+        raise RuntimeError(f"Failed to transition CSC to {self.desired_state!r}.")
+
+    async def _handle_csc_in_standby(self) -> bool:
+        """Handle condition where CSC is in STANDBY."""
+
+        try:
+            await self.scheduler_remote.cmd_start.set_start(
+                configurationOverride=self.configuration_override,
+                timeout=self.timeout_start,
+            )
+        except salobj.AckError:
+            self.log.warning("Sending start command failed. CSC not in STANDBY")
+            return False
+        else:
+            await self.set_desired_state()
+            return True
+
+    async def _handle_csc_in_disabled(self) -> bool:
+        """Handle condition where CSC is in DISABLED."""
+        try:
+            await self.scheduler_remote.cmd_standby.set_start(
+                timeout=self.timeout_start,
+            )
+        except salobj.AckError:
+            self.log.warning("Sending standby command failed. CSC not in DISABLED.")
+            return False
+        else:
+            await self.set_desired_state()
+            return True
+
+    async def _handle_csc_in_enabled(self) -> bool:
+        """Handler condition where CSC is in ENABLED."""
+        try:
+            await self.scheduler_remote.cmd_disable.set_start(
+                timeout=self.timeout_start,
+            )
+        except salobj.AckError:
+            self.log.warning("Sending disabled command failed. CSC not in ENABLED.")
+            return False
+        else:
+            await self.standby()
+            await self.set_desired_state()
+            return True
+
+    async def standby(self) -> None:
+        """Set Scheduler state to standby."""
+        await salobj.set_summary_state(
+            self.scheduler_remote,
+            salobj.State.STANDBY,
+        )
+
+    async def set_desired_state(self) -> None:
+        """Enable CSC."""
+        await salobj.set_summary_state(
+            self.scheduler_remote,
+            self.desired_state,
+            override=self.configuration_override,
+        )
+
+    async def run(self) -> None:
+        """Enable the Scheduler and exit."""
+
+        await self.checkpoint("Assert liveliness")
+        await self.assert_liveliness()
+
+        current_summary_state = self.get_summary_state()
+
+        if current_summary_state is None:
+            await self.checkpoint("Handling no summary state information")
+            await self.handle_no_summary_state_data()
```

Comment:
Well you warned me this would be controversial. This seems the wrong place to have such protections. if this is a common problem then perhaps I should add it to salobj set_summary_state.

For now, surely it would suffice to remove all this special handling of "no summary state" and just ask the observers to run the script a second time if it fails the first time?

Adding all this special code here just will make folks wonder why this one works sometimes when all the other scripts that try to set the state of a CSC do not.

Reply:
I wouldn't classify this as a "common problem" and I don't remember this ever affecting the Scheduler itself.. but, when this error happens users get extremely confused about what it going on and "running the script a second time" won't solve the problem as we still don't have historical data do make it work. 

Believe me, I dislike adding this code as much as you do, probably even more. But, I am trying to make the summit crew night time life easier with these scripts, minimizing the amount of typing required to operate the scheduler and anticipating some issues we've found from time to time. 

I honestly don't know if we should add this to salobj or not. Might be good, might be terrible, I simply don't know... As I said, I never wanted to add it in the first place, but I am fearful enough of this happening during the night that I did..

Reply:
If it hasn't affected Scheduler then this seems especially the wrong place for this code.

Reply:
well, until it happens in the middle of the night... I don't disagree with you.. I will probably remove it.. but it worries me much..

Reply:
ok, so I started removing this and I was going to add a message to help users work around the problem.. then it hit me, why I added this in the first place.. In order to recover from this problem if the CSC is in STANDBY, they will have to send the start command providing the configuration, which is quite an overhead/copy and pasting around.. more than they would normally do with other CSCs (which now won't require any configuration).. 

The scheduler will always require some configuration for operating for the perceivable future.. so, I think I'll keep this code and hopefully remove it once we either provide a workaround at a lower level or fix this historical data problem (Kafka?)...

Reply:
These two lines will always reconfigure the CSC, if the desired state is DISABLED or ENABLED:

await salobj.set_summary_state(..., state=salobj.State.STANDBY)
await salobj.set_summary_state(..., state=desired_state)